### PR TITLE
[Feat] support sub-partition and multi-column in key partition

### LINKED ISSUE #30: [Feat.]: support multi-column in key partition table
### Check Before Asking

- [X] Please check the issue list and confirm this feature is encountered for the first time.
- [X] Please try full text in English and attach precise description.

### Description

```markdown
support multi-column in the key partition table at the same level.
Support table like

# sql
CREATE TABLE `TEST_TABLE_PARTITION_COMPLEX_KEY` (
    `c1` bigint NOT NULL,
    `c2` varbinary(1024) NOT NULL,
    `c3` varchar(1024) NOT NULL,
    `c4` varchar(20) default NULL,
    PRIMARY KEY (`c1`, `c2`, `c3`)
) DEFAULT CHARSET = utf8mb4 ROW_FORMAT = DYNAMIC COMPRESSION = 'lz4_1.0' REPLICA_NUM = 3 BLOCK_SIZE = 16384 USE_BLOOM_FILTER = FALSE TABLET_SIZE = 134217728 PCTFREE = 10
partition by key(`c1`, `c2`, `c3`) partitions 16 ;
```

### Other Information

_No response_

### PR DESCRIPTION
**Pull request by IHEII**
### What changes were proposed in this pull request?

- Support sub-partition and multi-column in the key partition table, fix #30 and #31.

### Why are the changes needed?

- Support more functions in the key partition table

### Will break the compatibility? How if so?

N.A.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Pass all the tests in observer 3.x

### Checklist

- [x] I've run the tests to see all new and existing tests pass.
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above.
- [x] I've informed the technical writer about the documentation change if necessary.

### CODE REVIEW

File: src/client/table_client.rs [hunk 1]
```diff
@@ -761,9 +787,19 @@ impl ObTableClientInner {
...
+                            if first_part_len + sub_part_len > row_key.len() {
+                                return Err(CommonErr(
+                                    CommonErrCode::PartitionError,
+                                    "number of rowkey is less than the length of partition column"
+                                        .to_owned(),
+                                ));
+                            }
 
-                            let part_id2 = sub_part_desc.get_part_id(row_key);
+                            let part_id1 = first_part_desc.get_part_id(&row_key[..first_part_len]);
```
Comment (WeiXinChan) [inline src/client/table_client.rs:801]: part_id1 -> first_part_id
part_id2 -> sub_part_id

File: src/location/mod.rs [hunk 2]
```diff
@@ -387,8 +387,20 @@ impl ObPartitionEntry {
...
+        // logic_id = part_id in partition one
         self.parititon_location.get(&part_id)
     }
+
+    pub fn get_sub_partition_location_with_part_id(
+        &self,
+        part_id: i64,
+        sub_part_nums: i32,
+    ) -> Option<&ObPartitionLocation> {
+        // only for template sub partition
```
Comment (WeiXinChan) [inline src/location/mod.rs:399]: what mean "only for template sub partition"

Comment (IHEII) [reply src/location/mod.rs:399]: Template sub-partition means that this kind of partition could not add or delete a partition.
Observer supports other sub-partition methods that users could add partitions or reduce partitions, but obkv-client-rust doesn't support this feature yet.

File: src/location/ob_part_constants.rs [hunk 3]
```diff
@@ -15,6 +15,86 @@
...
+    }
+}
+
+// get subpart_id with PARTITION_LEVEL_TWO_MASK
+#[inline]
+pub fn extract_subpart_id(id: i64) -> i64 {
+    id & (!(u64::MAX << OB_PART_ID_SHIFT)) as i64
+}
+
+// get part_id with PARTITION_LEVEL_TWO_MASK
```
[the remark this answers is not in the document]
Comment (IHEII) [reply src/location/ob_part_constants.rs:38]: This function works like a mask, which means this function will get part of the data from the id.
btw, this comment is copied from observer orz.

File: src/client/table_client.rs (hunk 1 above)
Comment (IHEII) [reply src/client/table_client.rs]: fixed

**Comment by WeiXinChan** [comment]
LGTM

**Pull request merged**